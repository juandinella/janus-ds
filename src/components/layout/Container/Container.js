import isPropValid from '@emotion/is-prop-valid'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import styles from './Container.styles'

/**
 * Container
 */

const shouldForwardProp = (prop) => isPropValid(prop)
const StyledContainer = styled.div.withConfig({shouldForwardProp})`${styles}`


const Container = ({
  id,
  flex,
  space,
  mobileSpace,
  justifyContent,
  alignItems,
  alignContent,
  direction,
  isPlayground,
  wrap,
  overflow,
  tag,
  gap,
  mobileGap,
  children
}) => {
  return (
    <StyledContainer
      id={id}
      data-testid={id}
      flex={flex}
      space={space}
      mobileSpace={mobileSpace}
      justifyContent={justifyContent}
      alignItems={alignItems}
      alignContent={alignContent}
      direction={direction}
      wrap={wrap}
      overflow={overflow}
      as={tag}
      gap={gap}
      mobileGap={mobileGap}
      isPlayground={isPlayground}
    >
      {children}
    </StyledContainer>)
}

Container.displayName = 'Container'

Container.propTypes = {
  /** Container's id */
  id: PropTypes.string,
  /** Container's children */
  children: PropTypes.node.isRequired,
  /** This Boolean attribute add border to see it on Storybook */
  isPlayground: PropTypes.bool,
  /** Container Flexbox */
  flex: PropTypes.bool,
  /** Container's Padding */
  space: PropTypes.string,
  /** Container's Padding for mobile*/
  mobileSpace: PropTypes.string,
  /** The CSS align-items property sets the align-self value on all direct children as a group. In Flexbox, it controls the alignment of items on the Cross Axis. */
  alignItems: PropTypes.string,
  /** The CSS justify-content property defines how the browser distributes space between and around content items along the main-axis of a flex container. */
  justifyContent: PropTypes.string,
  /** The CSS align-content property sets the distribution of space between and around content items along a flexbox's cross-axis */
  alignContent: PropTypes.string,
  /** The flex-direction CSS property sets how flex items are placed in the flex container defining the main axis and the direction (normal or reversed). */
  direction: PropTypes.string,
  /** The flex-wrap CSS property sets whether flex items are forced onto one line or can wrap onto multiple lines. If wrapping is allowed, it sets the direction that lines are stacked. */
  wrap: PropTypes.string,
  /** Container's overflow */
  overflow: PropTypes.string,
  /** The space between containers */
  gap: PropTypes.string,
  /** The space between containers for mobile*/
  mobileGap: PropTypes.string,
}

Container.defaultProps = {
  tag: 'div',
}

export default Container
