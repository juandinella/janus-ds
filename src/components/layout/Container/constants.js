export const options = {
  /**
   * Justify Content
   * https://developer.mozilla.org/en-US/docs/Web/CSS/justify-content
   */
  justifyContent: [
    'flex-start',
    'center',
    'flex-end',
    'space-between',
    'space-around',
    'space-evenly',
    'stretch',
  ],
  /**
   * Align Items
   * https://developer.mozilla.org/en-US/docs/Web/CSS/align-items
   */
  alignItems: ['flex-start', 'center', 'flex-end', 'stretch', 'baseline'],
  /**
   * Align Content
   * https://developer.mozilla.org/en-US/docs/Web/CSS/align-content
   */
  alignContent: [
    'stretch',
    'center',
    'flex-start',
    'flex-end',
    'space-between',
    'space-around',
    'stretch',
  ],
  /**
   * Flex DIrection
   * https://developer.mozilla.org/en-US/docs/Web/CSS/flex-direction
   */
  direction: ['row', 'row-reverse', 'column', 'column-reverse'],
  /**
   * Flex Wrap
   * https://developer.mozilla.org/en-US/docs/Web/CSS/flex-wrap
   */
  wrap: ['nowrap', 'wrap', 'wrap-reverse'],
  spaces: ['0', 'xxs', 'xs', 'sm', 'md', 'lg', 'xl', 'xxl', 'xxxl'],
  borderRadiuses: ['xs', 'sm', 'lg'],
  borderColors: [
    'neutral-120',
    'neutral-100',
    'neutral-80',
    'neutral-60',
    'neutral-40',
    'neutral-20',
    'neutral-00',
    'primary-main',
  ],
  backgroundColors: [
    'neutral-120',
    'neutral-100',
    'neutral-80',
    'neutral-60',
    'neutral-40',
    'neutral-20',
    'neutral-00',
    'primary-main',
    'secondary-pink',
    'wallet1',
  ],
  overflow: ['hidden', 'visible', 'scroll', 'auto'],
  alignments: ['left', 'center', 'right'],
  whiteSpace: [
    'normal',
    'nowrap',
    'pre',
    'pre-wrap',
    'pre-line',
    'break-spaces',
  ],
}
