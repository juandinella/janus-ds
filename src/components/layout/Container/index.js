import Container from './Container'

export default Container
export { options } from './constants'
