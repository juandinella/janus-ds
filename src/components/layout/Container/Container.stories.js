import { options } from './constants'
import Container from './Container'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Container)
const ListTemplate = getListTemplate(Container)

export default {
  title: 'Layout/Container',
  component: Container,
  args: {
    children: 'This is a Child!',
    isPlayground: true,
    flex: true,
  },
  argTypes: {
    tag: {control: 'text'},
    gap: {
      control: 'text',
      description: 'The space between containers. Supports responsive design.',
    },
    mobileGap: {
      control: 'text',
      description: 'The space between containers for mobile.',
    },
    space: {
      control: 'text',
      description: 'Vertical and horizontal padding for desktop. E.g. "xs" or "xs, lg".'
    },
    mobileSpace: {
      control: 'text',
      description: 'Vertical and horizontal padding for mobile. If not provided, desktop values will be used.'
    },
    justifyContent: {
      ...getOptionsArgTypes(options.justifyContent, 'justifyContent'),
      description: 'Justify content for mobile and desktop, separated by a comma. E.g. "flex-start, center".'
    },
    alignItems: {
      ...getOptionsArgTypes(options.alignItems, 'alignItems'),
      description: 'Align items vertically in the container. E.g. "flex-start, center".',
    },
  },
}

export const Default = Template.bind({})

export const Responsive = { args: { space: 'xs, xl' } }

export const Spaces = ListTemplate.bind({})
Spaces.args = { items: options.spaces.map((space) => ({ space })) }
