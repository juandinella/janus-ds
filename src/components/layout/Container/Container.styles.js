import { css } from 'styled-components'
import layout from '../../../styles/theme/layout'

const parseTwoDimensionalSpacing = (spacing) => {
  const dimension1 = spacing && spacing.split(' ')[0]
  const dimension2 =
    (spacing && spacing.split(' ')[1]) ||
    (spacing && spacing.split(' ')[0])

  return [dimension1, dimension2]
}

export default css`
  width: 100%;
  position: relative;
  min-width: 1px;

  ${(props) =>
    props.flex &&
    css`
      display: flex;
      flex: 1;
    `}

  ${(props) =>
    props.isPlayground &&
    css`
      border: 2px dashed;
      border-color: var(--color-purple-600);
      min-height: 100px;
    `}



  ${(props) => {
    const [gapColumnMobile, gapRowMobile] = parseTwoDimensionalSpacing(props.mobileGap)
    const [gapColumn, gapRow] = parseTwoDimensionalSpacing(props.gap)

    return (
      props.gap &&
      css`
        gap: var(--spacing-${gapColumnMobile ? gapColumnMobile : gapColumn})
          var(--spacing-${gapRowMobile ? gapRowMobile : gapRow});

        @media screen and (min-width: ${layout.breakpoints.sm}) {
          gap: var(--spacing-${gapColumn}) var(--spacing-${gapRow});
        }
      `
    )
  }}

  ${(props) => {
    const [spaceColumnMobile, spaceRowMobile] = parseTwoDimensionalSpacing(props.mobileSpace)
    const [spaceColumn, spaceRow] = parseTwoDimensionalSpacing(props.space)

    return (
      props.space &&
      css`
        padding: var(--spacing-${spaceColumnMobile ? spaceColumnMobile : spaceColumn})
          var(--spacing-${spaceRowMobile ? spaceRowMobile : spaceRow});

        @media screen and (min-width: ${layout.breakpoints.sm}) {
          padding: var(--spacing-${spaceColumn}) var(--spacing-${spaceRow});
        }
      `
    )
  }}

  ${(props) => {
    if (props.justifyContent) {
      const [mobileJustifyContent, desktopJustifyContent] = props.justifyContent.split(",").map(s => s.trim())

      return css`
        justify-content: ${mobileJustifyContent};

        @media screen and (min-width: ${layout.breakpoints.sm}) {
          justify-content: ${desktopJustifyContent || mobileJustifyContent};
        }
      `
    }
  }}
`
