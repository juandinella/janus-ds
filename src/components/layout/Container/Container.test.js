import initStoryshots from '@storybook/addon-storyshots'
import { cleanup, render, screen } from '@testing-library/react'
import React from 'react'
import Container from './Container'

initStoryshots({
  storyKindRegex: /^Layout\/Container$/,
})

const componentProps = {
  id: 'container',
  children: <span data-testid="container_span">Im a span</span>,
}

afterEach(cleanup)

describe('<Container />', () => {
  test('should render component correctly', () => {
    render(<Container {...componentProps} />)
    const container = screen.getByTestId(componentProps.id)
    expect(container).toBeInTheDocument()
  })

  test('should render children correctly', () => {
    const { queryByTestId } = render(<Container {...componentProps} />)
    expect(queryByTestId(componentProps.id).children).toHaveLength(1)
    expect(queryByTestId('container_span')).toBeTruthy()
    expect(queryByTestId('container_span').textContent).toEqual('Im a span')
  })

  test('should render tag correctly', () => {
    render(<Container {...componentProps} tag="span" />)
    const container = screen.getByTestId(componentProps.id)
    expect(container.tagName.toLowerCase()).toBe('span')
  })
})
