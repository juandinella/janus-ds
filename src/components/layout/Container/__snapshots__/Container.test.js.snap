// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Layout/Container Default 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="sc-1kqcgfr-0 dgLeCe"
  >
    This is a Child!
  </div>
</div>
`;

exports[`Storyshots Layout/Container Responsive 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="sc-1kqcgfr-0 bjwHeg"
  >
    This is a Child!
  </div>
</div>
`;

exports[`Storyshots Layout/Container Spaces 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="sc-1kqcgfr-0 emMqYE"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 gNwBhk"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 fLfIsE"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 lmdbjI"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 jglFPI"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 jpdqlQ"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 bJZKRk"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 ilvRPc"
  >
    This is a Child!
  </div>
  <div
    className="sc-1kqcgfr-0 eecQGM"
  >
    This is a Child!
  </div>
</div>
`;
